kernel/src/table_features/column_mapping.rs

```diff
@@ -25,18 +29,135 @@ pub(crate) fn column_mapping_mode(
     protocol: &Protocol,
     table_properties: &TableProperties,
 ) -> ColumnMappingMode {
-    match table_properties.column_mapping_mode {
-        Some(mode) if protocol.min_reader_version() == 2 => mode,
-        Some(mode)
-            if protocol.min_reader_version() == 3
-                && protocol.has_reader_feature(&ReaderFeatures::ColumnMapping) =>
-        {
-            mode
-        }
+    match (
+        table_properties.column_mapping_mode,
+        protocol.min_reader_version(),
+    ) {
+        // NOTE: The table property is optional even when the feature is supported, and is allowed
+        // (but should be ignored) even when the feature is not supported. For details see
+        // https://github.com/delta-io/delta/blob/master/PROTOCOL.md#column-mapping
+        (Some(mode), 2) => mode,
+        (Some(mode), 3) if protocol.has_reader_feature(&ReaderFeatures::ColumnMapping) => mode,
         _ => ColumnMappingMode::None,
     }
 }
 
+/// When column mapping mode is enabled, verify that each field in the schema is annotated with a
+/// physical name and field_id; when not enabled, verify that no fields are annotated.
+pub fn validate_schema_column_mapping(schema: &Schema, mode: ColumnMappingMode) -> DeltaResult<()> {
+    if mode == ColumnMappingMode::Id {
+        // TODO: Support column mapping ID mode
+        return Err(Error::unsupported("Column mapping ID mode not supported"));
+    }
+
+    let mut validator = ValidateColumnMappings {
+        mode,
+        path: vec![],
+        err: None,
+    };
+    let _ = validator.transform_struct(schema);
+    match validator.err {
+        Some(err) => Err(err),
+        None => Ok(()),
+    }
+}
+
+struct ValidateColumnMappings<'a> {
+    mode: ColumnMappingMode,
+    path: Vec<&'a str>,
+    err: Option<Error>,
+}
+
+impl<'a> ValidateColumnMappings<'a> {
+    fn transform_inner_type(
+        &mut self,
+        data_type: &'a DataType,
+        name: &'a str,
+    ) -> Option<Cow<'a, DataType>> {
+        if self.err.is_none() {
+            self.path.push(name);
+            let _ = self.transform(data_type);
+            self.path.pop();
+        }
+        None
+    }
+    fn check_annotations(&mut self, field: &StructField) {
+        // The iterator yields `&&str` but `ColumnName::new` needs `&str`
+        let column_name = || ColumnName::new(self.path.iter().copied());
+        let annotation = "delta.columnMapping.physicalName";
```

Comment:
Is there any advantage to using `const` or `static` here? From this [source](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/const-and-static.html#:~:text=Rust%20provides%20a%20'global%20variable,a%20fixed%20location%20in%20memory.&text=Unlike%20let%20bindings%2C%20you%20must%20annotate%20the%20type%20of%20a%20static%20.): 
- Static lives in one place and is passed around by reference.
- Const is inlined wherever it is used.
- They don't mention let, but I imagine the compiler would behave like in the const case.

Perhaps this is too trivial to make a perf difference, but I was curious on your take @scovich

Reply:
All string literals are statically allocated. On *nix systems you can literally run the command `strings` on the compiled executable and it will print them out. The let/const we might write in code is just a `&'static` reference to that string constant.

Reply:
Aha right mb 😅  It's all in the data segment no matter what. Didn't know about the `strings` command, I just tried it and it's rly cool!